addons/account_edi_ubl/models/account_edi_format_en_16931.py

```diff
@@ -0,0 +1,243 @@
+
+    ####################################################
+    # Helpers
+    ####################################################
+
+    def _check_en_16931_invoice_configuration(self, invoice):
+        errors = []
+
+        # partners
+        customer = invoice.commercial_partner_id
+        if not self._get_ubl_partner_values(customer).get('en_16931_endpoint', False):
+            errors.append(_('The customer needs to have an endpoint id, typically this is the vat number. Please note that countries outside of EU are not supported.'))
+
+        supplier = invoice.company_id.partner_id.commercial_partner_id
+        if not self._get_ubl_partner_values(supplier).get('en_16931_endpoint', False):
+            errors.append(_('The supplier needs to have an endpoint id, typically this is the vat number. Please note that countries outside of EU are not supported.'))
+
+        # invoice line
+        if invoice.invoice_line_ids.filtered(lambda l: not (l.product_id.name or l.name)):
+            errors.append(_('Each invoice line must have a product or a label.'))
+
+        if invoice.invoice_line_ids.tax_ids.invoice_repartition_line_ids.filtered(lambda r: r.use_in_tax_closing) and \
+           not supplier.vat:
+            errors.append(_("When vat is present, the supplier must have a vat number."))
+
+        return errors
+
+    def _get_ubl_partner_values(self, partner):
+        # OVERRIDE
+        values = super()._get_ubl_partner_values(partner)
+        if partner.country_id.code in COUNTRY_EAS:
+            values['en_16931_endpoint'] = partner.vat
+            values['en_16931_endpoint_scheme'] = COUNTRY_EAS[partner.country_id.code]
+        return values
+
+    def _get_ubl_values(self, invoice):
+        ''' Get the necessary values to generate the XML. These values will be used in the qweb template when rendering.
+        Needed values differ depending on the implementation of the EN16931, as (sub)template can be overriden or called dynamically.
+        TO OVERRIDE
+
+        :returns:   a dictionary with the value used in the template has key and the value as value.
+        '''
+        self.ensure_one()
+
+        def convert_monetary(amount, from_currency):
+            # All monetary should be in the invoice currency, except for vat total
+            return from_currency._convert(amount, invoice.currency_id, invoice.company_id, invoice.invoice_date)
+
+        values = super()._get_ubl_values(invoice)
+        values.update({
+            'customization_id': 'urn:cen.eu:en16931:2017#compliant#urn:fdc:peppol.eu:2017:poacc:billing:3.0',
+            'profile_id': 'urn:fdc:peppol.eu:2017:poacc:billing:01:1.0',
+            'due_date': fields.Date.to_string(invoice.invoice_date_due),
+            'invoice_date': fields.Date.to_string(invoice.invoice_date),
+
+            'convert_monetary': convert_monetary,
+            'partner_template': 'account_edi_ubl.export_en_16931_invoice_partner',
```

Comment:
I don't think this dynamique template is necessary. If you want a custom one, you could inherit it or add your customization yourself using an `xpath`